### PR TITLE
VAULT-8631 - Change synchronous logic to insert storage entry after mounting

### PR DESCRIPTION
**Pull request by VioletHynes**
This is an addition to https://github.com/hashicorp/vault-plugin-secrets-kv/commit/cc06d01757ed8d20a9961be77bb46104c1be0805 as that pull request was not compatible with Vault making mounts read-only during mounting. Wasn't aware of this limitation - but this version works fully with Vault as far as I can tell.

The approach has been modified:
- We set to having upgraded immediately (which is what's used to determine if the mount is 'ready')
- We delay the insertion of the 'upgraded' entry in the mount until after it has been fully mounted

The two formatting changes made outside upgrade.go were done by `make fmt`.

### CODE REVIEW

File: upgrade.go [hunk 1]
```diff
@@ -201,6 +201,31 @@ func (b *versionedKVBackend) Upgrade(ctx context.Context, s logical.Storage) err
 		return nil
 	}
 
+	writeUpgradeInfoDoneFunc := func() {
+		upgradeInfo.Done = true
+		info, err := proto.Marshal(upgradeInfo)
+		if err != nil {
+			b.Logger().Error("encoding upgrade info resulted in an error", "error", err)
```
[the remark this answers is not in the document]
Comment (VioletHynes) [reply upgrade.go:208]: To what extent do we want to handle it in the main goroutine? Other than returning before the call to:
`atomic.StoreUint32(b.upgrading, 0)`
I'm not sure there's anything else - I feel like any logging etc should be in writeUpgradeInfoDoneFunc so it's shared between the sync/async logic. Is this the suggestion, or am I missing something?

I agree we shouldn't be trying to store incorrectly marshalled data, though. That should be an easy fix.

File: upgrade.go [hunk 2]
```diff
@@ -250,26 +275,16 @@ func (b *versionedKVBackend) Upgrade(ctx context.Context, s logical.Storage) err
...
-
-		err = s.Put(ctx, &logical.StorageEntry{
-			Key:   path.Join(b.storagePrefix, "upgrading"),
-			Value: info,
-		})
-		if err != nil {
-			b.Logger().Error("writing upgrade done resulted in an error", "error", err)
-		}
-
+		writeUpgradeInfoDoneFunc()
```
Comment (ccapurso) [inline upgrade.go:278]: Since we're calling this here do we still need the loop at the top of this func?

Comment (VioletHynes) [reply upgrade.go:278]: We do, yeah - the loop at the top is for the 'currently upgrading' entry, and this function adds the 'we're done upgrading' entry. We only write the second in the sync case, but writing both for a long upgrade is still relevant.

File: upgrade.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (ncabatoff) [reply upgrade.go:208]: Why not return an error to the user and fail the mount request?

File: upgrade.go (hunk 2 above)
Comment (ccapurso) [reply upgrade.go:278]: Ah write, I missed the difference in `Done` value. Whoops! Thanks for the clarification.

File: upgrade.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (ccapurso) [reply upgrade.go:208]: > Why not return an error to the user and fail the mount request?

I like this approach too

[the remark this answers is not in the document]
Comment (VioletHynes) [reply upgrade.go:208]: Updated! Thanks for the suggestion

File: upgrade.go [hunk 3]
```diff
@@ -201,6 +201,34 @@ func (b *versionedKVBackend) Upgrade(ctx context.Context, s logical.Storage) err
...
+	writeUpgradeInfoDoneFunc := func(info []byte) {
+	READONLY_LOOP:
+		for {
+			err = s.Put(ctx, &logical.StorageEntry{
+				Key:   path.Join(b.storagePrefix, "upgrading"),
+				Value: info,
+			})
+			switch {
+			case err == nil:
+				break READONLY_LOOP
```
Comment (ncabatoff) [inline upgrade.go:223]: You could get rid of the label and just return here instead.

File: upgrade.go [hunk 4]
```diff
@@ -201,6 +201,34 @@ func (b *versionedKVBackend) Upgrade(ctx context.Context, s logical.Storage) err
...
+			case err.Error() == logical.ErrSetupReadOnly.Error():
+				time.Sleep(10 * time.Millisecond)
+			default:
+				return
```
[the remark this answers is not in the document]
Comment (VioletHynes) [reply upgrade.go:227]: Oh, woops. Good catch - made a mistake between revisions

File: upgrade.go (hunk 3 above)
Comment (VioletHynes) [reply upgrade.go:223]: Good point - updating accordingly

File: upgrade.go [hunk 5]
```diff
@@ -212,18 +212,18 @@ func (b *versionedKVBackend) Upgrade(ctx context.Context, s logical.Storage) err
...
 				Value: info,
 			})
 			switch {
 			case err == nil:
-				break READONLY_LOOP
+				return
 			case err.Error() == logical.ErrSetupReadOnly.Error():
 				time.Sleep(10 * time.Millisecond)
 			default:
+				b.Logger().Error("writing upgrade info resulted in an error", "error", err)
```
Comment (ncabatoff) [inline upgrade.go:226]: I think this is a minor issue right?  Like next time a node becomes active it'll try this again, and there aren't any user-visible impacts other than the overhead of traversing the keys a second time to discover they're all already upgraded?  If so, we might want to provide some of that reassurance in the log message.

**Review by ncabatoff — APPROVED** [review]

File: upgrade.go (hunk 5 above)
Comment (VioletHynes) [reply upgrade.go]: Updated error message with the reassurance

**Pull request merged**